src/serac/physics/tests/serac_solid_functional_finite_diff.cpp

```diff
@@ -169,6 +169,146 @@ TEST(solid_functional_finite_diff, finite_difference)
+  double dt = 1.0;
+  solid_solver.advanceTimestep(dt);
+
+  // Output the sidre-based plot files
+  solid_solver.outputState();
+
+  // Make up an adjoint load which can also be viewed as a
+  // sensitivity of some qoi with respect to displacement
+  mfem::ParLinearForm adjoint_load_form(&solid_solver.displacement().space());
+  adjoint_load_form = 1.0;
+
+  // Construct a dummy adjoint load (this would come from a QOI downstream).
+  // This adjoint load is equivalent to a discrete L1 norm on the displacement.
+  serac::FiniteElementDual              adjoint_load(*mesh, solid_solver.displacement().space(), "adjoint_load");
+  std::unique_ptr<mfem::HypreParVector> assembled_vector(adjoint_load_form.ParallelAssemble());
+  adjoint_load.trueVec() = *assembled_vector;
+  adjoint_load.distributeSharedDofs();
+
+  // Solve the adjoint problem
+  solid_solver.solveAdjoint(adjoint_load);
+
+  // Compute the sensitivity (d QOI/ d state * d state/d parameter) given the current adjoint solution
+  [[maybe_unused]] auto& sensitivity = solid_solver.computeSensitivity<SHAPE_FIELD>();
+
+  // Perform finite difference on each shape velocity value
+  // to check if computed qoi sensitivity is consistent
+  // with finite difference on the displacement
+  double eps = 1.0e-6;
+  for (int i = 0; i < user_defined_shape_displacement.gridFunc().Size(); ++i) {
+    // Perturb the shape field
+    user_defined_shape_displacement.trueVec()(i) = shape_displacement_value + eps;
+    user_defined_shape_displacement.distributeSharedDofs();
+
+    solid_solver.advanceTimestep(dt);
+    mfem::ParGridFunction displacement_plus = solid_solver.displacement().gridFunc();
+
+    user_defined_shape_displacement.trueVec()(i) = shape_displacement_value - eps;
+    user_defined_shape_displacement.distributeSharedDofs();
+
+    solid_solver.advanceTimestep(dt);
+    mfem::ParGridFunction displacement_minus = solid_solver.displacement().gridFunc();
+
+    // Reset to the original bulk modulus value
+    user_defined_shape_displacement.trueVec()(i) = shape_displacement_value;
+    user_defined_shape_displacement.distributeSharedDofs();
+
+    // Finite difference to compute sensitivity of displacement with respect to bulk modulus
+    mfem::ParGridFunction ddisp_dshape(&solid_solver.displacement().space());
+    for (int i2 = 0; i2 < displacement_plus.Size(); ++i2) {
+      ddisp_dshape(i2) = (displacement_plus(i2) - displacement_minus(i2)) / (2.0 * eps);
+    }
+
+    // Compute numerical value of sensitivity of qoi with respect to bulk modulus
+    // by taking the inner product between adjoint load and displacement sensitivity
+    double dqoi_dshape = adjoint_load_form(ddisp_dshape);
+
+    // See if these are similar
+    SLIC_INFO(axom::fmt::format("dqoi_dshape: {}", dqoi_dshape));
+    SLIC_INFO(axom::fmt::format("sensitivity: {}", sensitivity.trueVec()(i)));
+    EXPECT_NEAR((sensitivity.trueVec()(i) - dqoi_dshape) / std::max(dqoi_dshape, 1.0e-2), 0.0, 5.0e-2);
```

Comment:
this tolerance seems really loose, can we be confident we're not making mistakes in the implementation?

Reply:
I think so - this QoI is extremely sensitive to nodal coordinate changes. It's probably worthwhile to come up with a more reasonable QoI for test.

Reply:
what is the QoI in this problem?

Reply:
I use a pseudo-load vector of 1 for all entries. This is equivalent to the L1 norm of the displacement vector (hence the large sensitivity).